x-pack/plugins/task_manager/server/task_claimers/strategy_mget.ts

```diff
@@ -319,69 +318,115 @@ async function searchAvailableTasks({
+      tasksWithPartitions(partitions)
+    );
+    searches.push({
+      query: queryUnlimitedTasks,
+      sort,
+      size,
+      seq_no_primary_term: true,
+    });
+  }
 
-  return await taskStore.fetch(
-    {
+  // add searches for limited types
+  for (const [type, capacity] of claimPartitions.limitedTypes) {
+    const queryForLimitedTasks = mustBeAllOf(
+      // Task must be enabled
+      EnabledTask,
+      // Specific task type
+      OneOfTaskTypes('task.taskType', [type]),
+      // Either a task with idle status and runAt <= now or
+      // status running or claiming with a retryAt <= now.
+      shouldBeOneOf(IdleTaskWithExpiredRunAt, RunningOrClaimingTaskWithExpiredRetryAt),
+      // must have a status that isn't 'unrecognized'
+      RecognizedTask
+    );
+
+    const query = matchesClauses(queryForLimitedTasks, filterDownBy(InactiveTasks));
+    searches.push({
       query,
       sort,
-      size,
+      size: capacity,
       seq_no_primary_term: true,
-    },
-    // limit the response size
-    true
-  );
+    });
+  }
+
+  return await taskStore.msearch(searches);
 }
 
-function applyLimitedConcurrency(
-  tasks: ConcreteTaskInstance[],
-  batches: TaskClaimingBatches
-): ConcreteTaskInstance[] {
-  // create a map of task type - concurrency
-  const limitedBatches = batches.filter(isLimited);
-  const limitedMap = new Map<string, number>();
-  for (const limitedBatch of limitedBatches) {
-    const { tasksTypes, concurrency } = limitedBatch;
-    limitedMap.set(tasksTypes, concurrency);
-  }
+interface ClaimPartitions {
+  removedTypes: string[];
+  unlimitedTypes: string[];
+  limitedTypes: Map<string, number>;
+}
 
-  // apply the limited concurrency
-  const result: ConcreteTaskInstance[] = [];
-  for (const task of tasks) {
-    const concurrency = limitedMap.get(task.taskType);
-    if (concurrency == null) {
-      result.push(task);
+interface BuildClaimPartitionsOpts {
+  types: Set<string>;
+  excludedTypes: Set<string>;
+  removedTypes: Set<string>;
+  getCapacity: (taskType?: string) => number;
+  definitions: TaskTypeDictionary;
+}
+
+function buildClaimPartitions(opts: BuildClaimPartitionsOpts): ClaimPartitions {
+  const result: ClaimPartitions = {
+    removedTypes: [],
+    unlimitedTypes: [],
+    limitedTypes: new Map(),
+  };
+
+  const { types, excludedTypes, removedTypes, getCapacity, definitions } = opts;
+  for (const type of types) {
+    if (excludedTypes.has(type)) continue;
+
+    if (removedTypes.has(type)) {
+      result.removedTypes.push(type);
       continue;
     }
 
-    if (concurrency > 0) {
-      result.push(task);
-      limitedMap.set(task.taskType, concurrency - 1);
+    const definition = definitions.get(type);
+    if (definition.maxConcurrency == null) {
+      result.unlimitedTypes.push(definition.type);
+      continue;
     }
+
+    const capacity = getCapacity(definition.type);
+    result.limitedTypes.set(definition.type, capacity);
```

Comment:
should we check for capacity=0 and not add to this map to avoid issuing a query with size 0?

Reply:
added in 842d4023a447ed91b81627faee668b0bcb2a1e16